extensions/smallrye-jwt/deployment/src/main/java/io/quarkus/smallrye/jwt/deployment/SmallRyeJwtProcessor.java

```diff
@@ -161,14 +168,31 @@ void registerOptionalClaimProducer(BeanRegistrationPhaseBuildItem beanRegistrati
                 continue;
             }
             AnnotationInstance claimQualifier = injectionPoint.getRequiredQualifier(CLAIM_NAME);
-            if (claimQualifier != null && injectionPoint.getType().name().equals(DotNames.PROVIDER)) {
-                // Classes from jakarta.json are handled specially
+            if (claimQualifier != null) {
+                boolean providerType = injectionPoint.getType().name().equals(DotNames.PROVIDER);
+                boolean instanceType = injectionPoint.getType().name().equals(DotNames.INSTANCE);
+
                 Type actualType = injectionPoint.getRequiredType();
-                if (actualType.name().equals(DotNames.OPTIONAL) && !actualType.name().toString()
-                        .startsWith("jakarta.json")) {
+
+                Optional<BeanInfo> bean = injectionPoint.getTargetBean();
+                if (bean.isPresent()) {
+                    DotName scope = bean.get().getScope().getDotName();
+                    if (!REQUEST_SCOPED_NAME.equals(scope)
+                            && (!providerType && !instanceType && !CLAIM_VALUE_NAME.equals(actualType.name()))) {
```

Comment:
In theory, a user could also inject `io.quarkus.arc.InjectableInstance` for programmatic lookup.